### ISSUE TITLE
M-1: Forfeited junior loss recovery lock — admin ATA token trap in ReturnInsurance

### ISSUE THREAD

**Issue by v1ktorrr0x** [dcccrypto/percolator-stake#197]
## Bug

When the last junior depositor exits during an active insurance loss, `process_withdraw` increments `pool.total_returned` by the forfeited amount `L` without any physical token return:

```rust
// process_withdraw — last-junior exit block
pool.total_returned = pool
    .total_returned
    .checked_add(forfeited)
    .ok_or(StakeError::Overflow)?;
```

`process_return_insurance` then caps any return to:

```rust
let outstanding = pool.total_flushed.saturating_sub(pool.total_returned);
if amount > outstanding {
    return Err(ProgramError::InvalidArgument);
}
```

Because `total_returned` was raised by `L` without a matching physical deposit, `outstanding` is reduced by `L`. When the wrapper resolves the claim and physically returns the full flushed amount to the admin, the admin's `ReturnInsurance` call will exceed the artificially reduced `outstanding` limit and revert with `InvalidArgument`.

## Impact

Admin cannot return the full physically recovered insurance amount to the vault. The vault ends up short by `L` tokens, causing subsequent senior withdrawals to fail with `InsufficientVaultBalance` unless the admin manually bypasses the program to top up the vault.

## Fix

Bound `amount` against the physical outstanding (including forfeited junior loss), and only increment `total_returned` up to the active portion:

```rust
// Validate against physical outstanding
let outstanding_physical = pool
    .total_flushed
    .checked_add(pool.realized_junior_loss())
    .ok_or(StakeError::Overflow)?
    .saturating_sub(pool.total_returned);
if amount > outstanding_physical {
    return Err(ProgramError::InvalidArgument);
}

// Only raise total_returned up to the active (non-forfeited) outstanding
let returned_increment = amount.min(
    pool.total_flushed.saturating_sub(pool.total_returned)
);
pool.total_returned = pool
    .total_returned
    .checked_add(returned_increment)
    .ok_or(StakeError::Overflow)?;
```

Any excess above `returned_increment` represents the forfeited junior loss being physically refilled — transfer those tokens to the vault as dead value without touching `total_returned`.

**Comment by 0x-SquidSol** [comment]
Thank you for the detailed write-up. We traced this end-to-end against current `main` and it does **not** reproduce: the described impact (vault left short by `L`, senior withdrawals failing, admin forced to bypass the program) cannot occur. The behavior described is the intended #161 fair-recovery design, and a conservation invariant guarantees the vault always backs senior's claim exactly.

## Why there is no shortfall

The last-junior-exit booking increments **both** `total_returned` **and** `realized_junior_loss` by the same `L` (processor.rs, last-junior block):

```rust
pool.total_returned = pool.total_returned.checked_add(forfeited)?;        // +L
pool.set_realized_junior_loss(pool.realized_junior_loss().checked_add(forfeited)?); // +L
```

`total_pool_value()` subtracts `realized_junior_loss`:

```
total_pool_value = deposited − withdrawn − flushed + returned + fees − realized_junior_loss
```

so the two `+L` terms cancel — the booking is **value-neutral** for senior. The key point: the `+L` added to `total_returned` is a *phantom* return (no tokens move at exit); it is recorded precisely in `realized_junior_loss`. Therefore the physically-returned tokens are `total_returned − realized_junior_loss`, and the physical vault balance is:

```
physical_vault = deposited − withdrawn − flushed + (returned − realized_junior_loss)
              = total_pool_value()
```

This holds at **every** step. The vault is never short of senior's claim. The reduced `outstanding = total_flushed − total_returned` cap is the design (#161): it stops a later recovery of the *junior's* forfeited loss from being re-credited as a *senior* windfall (senior was protected and must not profit from the junior's loss).

## What actually happens

Take senior 100, junior 50, flush 120 (the loss spills past junior into senior), the last junior exits, then the wrapper resolves and returns the full 120 to the admin's ATA:

- After the forfeit booking, senior's claim is unchanged (30) and the vault physically holds 30 — no shortfall.
- `outstanding = 120 − 50 = 70`. The admin returns **70** (the senior-recoverable portion), which restores senior to its full 100 with the vault holding exactly 100. Senior withdraws in full.
- Attempting to return the full 120 returns `InvalidArgument` — this is **correct over-return rejection**, not a trap. The remaining 50 is the junior's forfeited loss and is correctly excluded from senior's claim.

Two incidental notes: `InsufficientVaultBalance` is raised only on the **flush** path, never on withdraw; and the proposed fix would merely move the intentionally-dead `L` from the admin's ATA into the vault as unaccounted tokens — a cosmetic relocation that fixes no reachable shortfall.

## Reproduction (disproof)

This test drives the real `StakePool` accounting methods through flush → loss spilling into senior → last-junior forfeit booking → wrapper recovery → `ReturnInsurance(outstanding)` → full senior withdrawal, asserting `physical_vault == total_pool_value()` at every step:

```rust
use bytemuck::Zeroable;
use percolator_stake::state::StakePool;

fn pool() -> StakePool {
    let mut p = StakePool::zeroed();
    p.is_initialized = 1;
    p.set_discriminator();
    p.set_tranche_enabled(true);
    p
}

// Physical tokens in the vault: the #161 booking raises total_returned by L WITHOUT a
// token movement and records that phantom L in realized_junior_loss, so physically
// returned tokens == total_returned - realized_junior_loss.
fn physical_vault(p: &StakePool) -> i128 {
    p.total_deposited as i128 - p.total_withdrawn as i128 - p.total_flushed as i128
        + (p.total_returned as i128 - p.realized_junior_loss() as i128)
}

fn assert_vault_backs_claim(p: &StakePool, ctx: &str) {
    let tpv = p.total_pool_value().unwrap();
    assert_eq!(physical_vault(p), tpv as i128, "physical vault must equal total_pool_value ({ctx})");
}

#[test]
fn forfeit_booking_never_leaves_vault_short_of_senior() {
    let mut pool = pool();

    // Senior 100, Junior 50. Vault physically holds 150.
    pool.total_deposited = 150;
    pool.total_lp_supply = 150;
    pool.set_junior_balance(50);
    pool.set_junior_total_lp(50);
    assert_eq!(pool.senior_total_lp(), 100);
    assert_vault_backs_claim(&pool, "genesis");

    // Flush 120. Loss spills past junior (50) into senior (70).
    pool.total_flushed = 120;
    assert_eq!(pool.effective_junior_balance(), 0);
    assert_eq!(pool.senior_balance().unwrap(), 30);
    assert_vault_backs_claim(&pool, "after loss, pre-exit");

    // Last junior exits (payout 0) — replicate the #161 booking exactly.
    let net_loss = pool.total_flushed.saturating_sub(pool.total_returned); // 120
    let forfeited = pool.junior_balance()
        .saturating_sub(pool.effective_junior_balance())
        .min(net_loss); // 50
    pool.total_returned += forfeited;                                       // phantom
    pool.set_realized_junior_loss(pool.realized_junior_loss() + forfeited);
    pool.set_junior_balance(0);
    pool.set_junior_total_lp(0);
    pool.total_lp_supply -= 50;
    assert_eq!(pool.total_pool_value().unwrap(), 30, "#161 booking is value-neutral");
    assert_vault_backs_claim(&pool, "after last-junior forfeit booking");

    // Wrapper returns the full 120; admin may return only the active outstanding.
    let outstanding = pool.total_flushed.saturating_sub(pool.total_returned); // 70
    assert_eq!(outstanding, 70);
    pool.total_returned += outstanding;                                      // physical 70 -> vault

    // Senior fully restored; vault holds exactly 100; no shortfall.
    assert_eq!(pool.senior_balance().unwrap(), 100);
    assert_vault_backs_claim(&pool, "after ReturnInsurance(outstanding)");
    assert!(physical_vault(&pool) >= pool.senior_balance().unwrap() as i128);
}
```

```
running 2 tests
test forfeit_booking_never_leaves_vault_short_of_senior ... ok
test forfeit_booking_value_neutral_when_junior_fully_absorbs ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

## Disposition

No vault shortfall and no senior-withdrawal failure exists on current `main`; the conservation invariant `physical_vault == total_pool_value()` holds throughout, and the `outstanding` cap behaves as designed (#161). **This issue can be closed.**

(For completeness: when a previously-forfeited junior loss is later recovered by the wrapper, those `L` tokens land in the admin's ATA rather than being re-credited to senior — a deliberate dead-value/windfall-allocation choice, not the shortfall described here. If sweeping that recovered remainder somewhere specific is ever desired, that is a separate design discussion.)

**Comment by dcccrypto** [comment]
Verified NOT_REAL against current main. The reported recovery-lock is value-neutral: the #161 last-junior-exit booking does `total_returned += L` AND `realized_junior_loss += L`, and total_pool_value() subtracts realized_junior_loss — so the bump is fully offset and senior is never windfalled or bricked. The vault fully backs senior before any ReturnInsurance; the outstanding-cap is intended #161 behavior (recovered L is documented dead value). Proven by test_161_last_junior_exit_does_not_windfall_recovery_to_senior. Closing.
